common/GL/ShaderCache.cpp

```diff
@@ -0,0 +1,402 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2021 PCSX2 Dev Team
+ *
+ *  PCSX2 is free software: you can redistribute it and/or modify it under the terms
+ *  of the GNU Lesser General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  PCSX2 is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with PCSX2.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "common/GL/ShaderCache.h"
+#include "common/FileSystem.h"
+#include "common/Console.h"
+#include "common/MD5Digest.h"
+#include "common/StringUtil.h"
+
+namespace GL
+{
+#pragma pack(push, 1)
+	struct CacheIndexEntry
+	{
+		u64 vertex_source_hash_low;
+		u64 vertex_source_hash_high;
+		u32 vertex_source_length;
+		u64 geometry_source_hash_low;
+		u64 geometry_source_hash_high;
+		u32 geometry_source_length;
+		u64 fragment_source_hash_low;
+		u64 fragment_source_hash_high;
+		u32 fragment_source_length;
+		u32 file_offset;
+		u32 blob_size;
+		u32 blob_format;
+	};
+#pragma pack(pop)
+
+	ShaderCache::ShaderCache() = default;
+
+	ShaderCache::~ShaderCache()
+	{
+		Close();
+	}
+
+	bool ShaderCache::CacheIndexKey::operator==(const CacheIndexKey& key) const
+	{
+		return (
+			vertex_source_hash_low == key.vertex_source_hash_low && vertex_source_hash_high == key.vertex_source_hash_high &&
+			vertex_source_length == key.vertex_source_length && geometry_source_hash_low == key.geometry_source_hash_low &&
+			geometry_source_hash_high == key.geometry_source_hash_high &&
+			geometry_source_length == key.geometry_source_length && fragment_source_hash_low == key.fragment_source_hash_low &&
+			fragment_source_hash_high == key.fragment_source_hash_high && fragment_source_length == key.fragment_source_length);
+	}
+
+	bool ShaderCache::CacheIndexKey::operator!=(const CacheIndexKey& key) const
```

Comment:
Protip - defining this operator as `return !(this == key);` allows for much less typing. It doesn't matter now since it's typed out as-is but it's worth mentioning IMO